File: dfa.hs
-- $ ghc --make dfa
-- $ ./dfa
-- a => False
-- baa => False
-- baba => True

import Data.List

data FARule = FARule Integer Char Integer deriving Show
data DFARulebook = DFARulebook [FARule] deriving Show
data DFA = DFA Integer [Integer] DFARulebook deriving Show

appliesTo (FARule s c n) s' c' = s == s' && c == c'
follow (FARule s c n) = n

ruleFor (DFARulebook (r:rs)) s' c' = if (appliesTo r s' c') then r else ruleFor (DFARulebook rs) s' c'
nextState rulebook s' c' = follow $ ruleFor rulebook s' c' 

isAccepting (DFA s as _) = s `elem` as 
readCharacter (DFA s as rulebook) c' = DFA (nextState rulebook s c') as rulebook
readString dfa ss = foldl readCharacter dfa ss

isAccepts dfa ss = isAccepting $ readString dfa ss

rules = [ FARule 1 'a' 2, FARule 1 'b' 1,
          FARule 2 'a' 2, FARule 2 'b' 3,
          FARule 3 'a' 3, FARule 3 'b' 3 ]
rulebook = DFARulebook rules

test s =
  putStrLn $ s ++ " => " ++ (show $ isAccepts (DFA 1 [3] rulebook) s)

main = do
  test "a"
  test "baa"
  test "baba"

File: dfa.erl
% $ erlc dfa.erl
% $ erl -noshell -s dfa start -s init stop
% a => false
% baa => false
% baba => true

-module(dfa).
-export([start/0]).

appliesTo({rule, State0, Character0, _}, State1, Character1) ->
  (State0 == State1) and (Character0 == Character1).

follow({rule, _, _, NextState}) ->
  NextState.

ruleFor([Rule|Rules], State, Character) ->
  case appliesTo(Rule, State, Character) of
    true -> Rule;
    false -> ruleFor(Rules, State, Character)
  end.

nextState(Rules, State, Character) ->
  Rule = ruleFor(Rules, State, Character),
  follow(Rule).

isAccepting({dfa, State, AcceptStates, _}) ->
  lists:member(State, AcceptStates).

readCharacter(DFA, Character) ->
  {dfa, State0, AcceptStates, Rules} = DFA,
  State1 = nextState(Rules, State0, Character),
  {dfa, State1, AcceptStates, Rules}.

readString(DFA, "") ->
  DFA;
readString(DFA, [C|CS]) ->
  DFA2 = readCharacter(DFA, C),
  readString(DFA2, CS).

isAccepts(DFA, String) ->
  DFA1 = readString(DFA, String),
  isAccepting(DFA1).

rules() -> [
  {rule, 1, $a, 2}, {rule, 1, $b, 1},
  {rule, 2, $a, 2}, {rule, 2, $b, 3},
  {rule, 3, $a, 3}, {rule, 3, $b, 3}
].

test(DFA, String) ->
  io:format("~s => ~p~n", [String, isAccepts(DFA, String)]).

start() ->
  Rules = rules(),
  DFA = {dfa, 1, [3], Rules},
  test(DFA, "a"),
  test(DFA, "baa"),
  test(DFA, "baba").
